Introduce network abstraction for butterfly

krnowak opened the pull request:
This PR adds a network abstraction to be used by butterfly server and client. The reason for introducing it is to be able to add a different implementation of the abstraction for testing purposes. That testing implementation should make mocking some scenarios way easier than it is currently possible. One of those scenarios would be talking to the butterfly server which is behind a NAT.

The PR also contains some cleanups I had to make for rustfmt to stop complaining.

Commit:
pkg-export-k8s: Add 2 simple unit tests to Manifest

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to ManifestJson

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Bind

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Manifest

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Use str::parse()

Instead of more explicit FromStr::from_str() to create a Topology from a
string.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: testcase for natural num validator

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

krnowak committed:
Use the new network abstraction in the butterfly server

This commit makes the `Server` type to be parameterized with a type
implementing the `Network` trait (in rust, it changes the type from
'struct Server` to `struct Server<N: Network>`). This change also
cascades to several other structs which contain the server, namely to
`Expire`, `Pull`, `Push`, `Inbound` and `Outbound` structs. The
functions that were taking a server as a parameter are also affected.

On the other hand, the tests are currently hardcoded to use the
concrete implementation of the `Network` trait.

Supervisor uses the butterfly server with the `RealNetwork`
implementation, of course.

Note that the `Server::new` function used to take two parameters of
types implementing the `ToSocketAddrs` trait. These were replaced with
one parameter of type that implements the `Network` trait. This means
that the `Server::new` function is not doing the name resolving
anymore. However, the name resolving functionality was not used
anywhere - all the uses of the `Server::new` function were passing
`SocketAddr` instances to it and `SocketAddr` implements the
`ToSocketAddrs` trait basically with an identity function.

With the name resolving functionality being gone, the `Server::new`
function cannot fail anymore.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Fix some rustfmt issues

Mostly too long lines on which rustfmt was bailing out in the files I
modified in previous commits.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Use the new network abstraction in butterfly client

This commit makes the `Client` type to be parameterized with a type
implementing the `GossipSender` trait (in rust, this changes the type
from `struct Client` to `struct Client<GS: GossipSender>`).

This also changes the prototype of the `Client::new` function, so now
it takes a socket instead of creating it and thus it is not possible
for this function to fail anymore.

The tests are hardcoded to use the `RealNetwork` and
`GossipZmqSocket`, but to avoid repetitive code, some helper functions
for creating a `Client` instance were added.

This change also affects the "hab config" utility, so it also uses the
`RealNetwork` and `GossipZmqSocket`.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Drop unused code and deps

Nothing uses ZMQ_CONTEXT anymore, which was also the only user of
lazy_static package. Drop them.

Also move some code to the only place it is used.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

lilic commented on components/butterfly-test/src/lib.rs:
```diff
@@ -398,12 +435,28 @@ impl SwimNet {
+                    trace_it!(
+                        TEST: &self.members[from_entry],
+                        format!(
+                            "Health {} {} as {}",
```
Maybe it should be `Health failed {} ...` as the examples below, or maybe I am misunderstanding?

krnowak replied:
Oh, I didn't add this. The change here is only because rustfmt complained about too long lines.

lilic replied:
Yes, I understand but maybe it should be added? :)

krnowak replied:
I don't think so. This `trace_it` macro is executed when `real_health == health` so apparently nothing failed here and the function returns true, so the `trace_it` macro below is executed when apparently `real_health != health` which means that something has indeed failed.

I wouldn't change it anyway - this code is not relevant at all to my changes. :P

Commit:
pkg-export-k8s: Document export_for_cli_matches() function

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Add docs to Cli

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

asymmetric commented on components/butterfly/src/error.rs:
```diff
@@ -71,6 +77,15 @@ impl fmt::Display for Error {
                     err
                 )
             }
+            Error::GossipPullSocketReceiveError(ref str) => {
```
Seems like most other cases use `err` as a variable name.

krnowak replied:
Alright.

asymmetric commented on components/butterfly/src/network.rs:
```diff
@@ -0,0 +1,297 @@
+}
+
+/// A trait for types used for sending gossip messages.
+pub trait GossipPushSocket {
```
Isn't "socket" too specific a term? Could there be other implementations that are not sockets? In that case, something like "GossipPusher" sounds more appropriate for a trait to me.

krnowak replied:
Maybe "GossipSender" and "GossipListener"?

asymmetric commented on components/butterfly/src/network.rs:
```diff
@@ -0,0 +1,297 @@
+
+/// A trait for types used for sending gossip messages.
+pub trait GossipPushSocket {
+    /// Send bytes being a gossip message.
```
I don't understand this sentence.

krnowak replied:
Abracadabra, hocus pocus, what's here not to understand? ;)

Right, it's terrible. Will reword it.

asymmetric commented on components/butterfly/src/network.rs:
```diff
@@ -0,0 +1,297 @@
+// on). I added it here, because Network is exclusively used by the
+// butterfly component.
+//
+// Same for Debug - Network is used by the butterfly component only so
```
Debug I can totally agree on, as it makes debugging easier. No idea about `'static`. What happens if you remove it?

krnowak replied:
It doesn't compile. :)

IIRC, rustc complains something about type (sic!) not living long enough or some gobbledegook.

I'll remove it and paste the complaints.

```
error[E0310]: the parameter type `N` may not live long enough
   --> src/server/mod.rs:258:14
    |
126 | impl<N: Network> Server<N> {
    |      -- help: consider adding an explicit lifetime bound `N: 'static`...
...
258 |             .spawn(move || {
    |              ^^^^^
    |
note: ...so that the type `[closure@src/server/mod.rs:258:20: 261:14 server_a:server::Server<N>, socket_a:<N as network::Network>::SwimSocket, tx_outbound:std::sync::mpsc::Sender<(std::net::SocketAddr, message::swim::Swim)>]` will meet its required lifetime bounds
   --> src/server/mod.rs:258:14
    |
258 |             .spawn(move || {
    |              ^^^^^

error[E0310]: the associated type `<N as network::Network>::SwimSocket` may not live long enough
   --> src/server/mod.rs:258:14
    |
258 |             .spawn(move || {
    |              ^^^^^
    |
    = help: consider adding an explicit lifetime bound `<N as network::Network>::SwimSocket: 'static`...
note: ...so that the type `[closure@src/server/mod.rs:258:20: 261:14 server_a:server::Server<N>, socket_a:<N as network::Network>::SwimSocket, tx_outbound:std::sync::mpsc::Sender<(std::net::SocketAddr, message::swim::Swim)>]` will meet its required lifetime bounds
   --> src/server/mod.rs:258:14
    |
258 |             .spawn(move || {
    |              ^^^^^

```

asymmetric commented on components/butterfly/src/error.rs:
```diff
@@ -33,6 +33,9 @@ pub enum Error {
     BadMessage(String),
     CannotBind(io::Error),
     DatFileIO(PathBuf, io::Error),
+    GossipPullSocketReceiveError(String),
```
Hadn't we agreed on using the failure crate? How does this play into that?

/cc @zeenix

krnowak replied:
I would need to analyze first, how the errors are used in butterfly and convert the whole component to it first. There are several ways of using failure, so I stated I won't waste time on it.

asymmetric commented on components/butterfly/src/network.rs:
```diff
@@ -0,0 +1,297 @@
+    /// For getting swim sockets, get an instance with
+    /// new_for_server().
+    pub fn new_for_client() -> Self {
+        Self::new(Self::throw_away_addr(), Self::throw_away_addr(), -1)
```
Why does it use "throw-away" addresses? Maybe a comment could explain that.

krnowak replied:
Client only sends gossip messages (rumors) and the target of the message is specified in the get_gossip_push_socket function. The addresses here are for receiving/sending SWIM messages or receiving rumors, which client does not do.

Will add an explanation there.

asymmetric commented on components/butterfly/src/member.rs:
```diff
@@ -131,14 +131,28 @@ pub struct Member {
+    /// # Panics
+    ///
+    /// This function panics if the address is un-parseable. In practice, it shouldn't be
+    /// un-parseable, since its set from the inbound socket directly.
```
Parsable is the word you're looking for I think ;) (at least according to [dictionary.com](http://www.dictionary.com/browse/parsable))

krnowak replied:
Heh, it's a copy-pasta from the function above it. Will fix both.

asymmetric commented on components/butterfly-test/src/lib.rs:
```diff
@@ -20,6 +20,7 @@ extern crate time;
 extern crate habitat_butterfly;
```
Don't know how to comment on that but the commit message went way over my head (The first sentence especially)

Is there a way to reformulate it without using the forbidden c-word ("class")?

krnowak replied:
Yeah, sounds crappy. Probably need to figure out the proper rustacean wording here. And probably elaborate a bit more.

asymmetric commented on components/butterfly/src/network.rs:
```diff
@@ -134,6 +132,19 @@ impl GossipPullSocket for ZmqGossipSocket {
+
+impl ServerContext {
+    pub fn as_mut(&self) -> &mut zmq::Context {
+        unsafe { &mut *self.0.get() }
```
Why is the `unsafe` needed here? Should it be documented in some way?

krnowak replied:
No idea, I didn't write it - I moved this code to the only file where it's used.

asymmetric reviewed (changes requested):
As a general feedback, I found the PR hard to review, since it's so long (and complex). I think splitting commits might've helped.

krnowak commented:
About the size of the commits - I tried to make them small, but still buildable. Maybe I'll try to explain better the changes I have made in the commit message, so it will be clear why they are here and why so many.

Commit:
pkg-export-k8s: Require base64 crate

In a following patch, we'll use this to encode configuration file content
into base64.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: user-config-secret -> config

The handling of the configuration file was all wrong: The part of the
manifest that is supposed to create the configuration file, was always
commented out and instead of taking the configuration file from the user,
we were expecting them to provide the name for the file. This patch fixes
this issue.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Add a doc comment to Topology enum

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Slightly more thorough Topology tests

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Fix formatting to satisfy rustfmt

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Ditch custom Topology type

This type with the same API and tests already exists in the habitat_sup
crate so let's just make use of that and ditch our version.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Ditch custom Bind type

This type with the same API and tests already exists in the habitat_sup
crate so let's just make use of that and ditch our version.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Drop a redundant 'mut'

One less compiler warning.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
sup: Make use of the result

Not only this is the right thing to do, it also get rid of compiler
warnings about this.

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

Commit:
pkg-export-k8s: Fix a minor grammatical mistake

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

mgamini committed:
adding Getting Started with Node blog post

Signed-off-by: Garrett Amini <kadjar@gmail.com>

adding author

Signed-off-by: Garrett Amini <kadjar@gmail.com>

Commit:
Add missing commas in .bldr.toml

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

krnowak committed:
Add network abstraction for butterfly

This commit adds a couple of traits as an abstraction for
networking. The abstraction is not trying to be very generic - it is
tailored for the butterfly server and client needs, so it may be a bit
leaky (we are using `SocketAddr` instead of providing yet another
trait that would address the addressing), but it's fine for its
intended purposes.

The reason for introducing the abstraction is to be able to use a
different implementation for testing needs. For example to mock a
scenario which may be hard to simulate with UDP and ZMQ sockets on
localhost. An example of the scenario would be talking to the
butterfly server behind a NAT.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

Commit:
pkg-export-k8s: Explicit ":latest" tag on docker image

Signed-off-by: Zeeshan Ali <zeeshan@kinvolk.io>

krnowak committed:
Add a network implementation for real-world usage

The implementation of the network abstraction uses UDP and ZeroMQ
sockets for SWIM and gossip communication, respectively, just like the
butterfly server and client currently do. This implementation will be
used by Supervisor and by "hab config" in the following commits.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Change "un-parseable" to "unparsable" in function docs

And wrap the docs a bit.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Avoid the "socket" terms in members, variables and types

Rename them to a bit more specific terms (which can describe their
function too), because the networking was abstracted from sockets into
SWIM channels, gossip senders and receivers.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

raskchanky committed:
Removing builder from the main repo

Signed-off-by: Josh Black <raskchanky@gmail.com>

krnowak committed:
Drop unused errors

Some of them were unused even without my changes.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

krnowak committed:
Drop a file with unused code

The file was confusing as the lot of code there was very similar to
the code in lib.rs, but with some small differences.

Also, the net.rs file was not mentioned in lib.rs, means it wasn't
even built.

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

christophermaier committed:
Remove some crates from the Makefile

`core` and `net` are no longer in this repository.

christophermaier committed:
Update habitat_core dependency

Point to the `master` branch again.

christophermaier committed:
Drop handlebars dependency back to pre-split level

Ran `cargo update --package=handlebars --precise=0.29.1`

A dependency of `handlebars` (`pest`) wasn't able to compile for some
reason. Until that can be sorted out, using the same versions that we
had before we split the repository apart seems sensible.

Signed-off-by: Christopher Maier <cmaier@chef.io>

christophermaier committed:
No need to build core on AppVeyor anymore!

Signed-off-by: Christopher Maier <cmaier@chef.io>

mwrock committed:
adding builder-depot-client compilation back to appveyor and makefile

Signed-off-by: mwrock <matt@mattwrock.com>

thesentinels committed:
Merge pull request #4452 from habitat-sh/jb/remove-builder

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

Commit:
Remove files that were used exclusively to run builder

Signed-off-by: Travis Elliott Davis <edavis@chef.io>

thesentinels committed:
Merge pull request #4489 from habitat-sh/elliott/builder_cleanup

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #4464 from habitat-sh/blog/getting-started-node

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #4448 from kinvolk/zeenix/k8s-tests

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #4481 from kinvolk/krnowak/drop-old-file

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

krnowak commented:
Filed the PR upstream: https://github.com/habitat-sh/habitat/pull/4497
